warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,371 @@
+   * @param proximityZoneMaxSpeed       Maximum instant speed to consider you are in the same proximity zone
+   * @param proximityInZoneMaxMeanSpeed Minimum mean speed in the proximityZone to prevent a split, in m/s
+   * @param splitTypeLabel              If set, add a label with the reason of the split
+   * @param stoppedTimeLabel            If set, add a label with the time in the stop state
    * @return
    * @throws WarpScriptException
    */
-  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String label, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneSpeed, long proximityZoneTime) throws WarpScriptException {
+  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String splitNumberLabel, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneMaxSpeed,
+                                                long proximityZoneTime, double proximityInZoneMaxMeanSpeed, String splitTypeLabel, String stoppedTimeLabel, boolean proximityZoneSplit) throws WarpScriptException {
     //
     // Sort GTS according to timestamps
     //
-    
+
     GTSHelper.sort(gts, false);
 
     GeoTimeSerie split = null;
-    
+
     List<GeoTimeSerie> splits = new ArrayList<GeoTimeSerie>();
-    
+
     int idx = 0;
-    
+    int gtsid = 1;
     int n = GTSHelper.nvalues(gts);
-    
+
     boolean mustSplit = true;
 
     long refLocation = GeoTimeSerie.NO_LOCATION;
     long refTimestamp = Long.MIN_VALUE;
-    
+
     long previousTimestamp = Long.MIN_VALUE;
     long previousLocation = GeoTimeSerie.NO_LOCATION;
     double proximityZoneTraveledDistance = 0.0D;
-    
+    String splitReason = SPLIT_TYPE_END;
+
+    //
+    // empty gts, or gts with only one value, returns a clone + extra labels when set
+    //
+    if (n <= 1) {
+      GeoTimeSerie result = gts.clone();
+      if (null != splitTypeLabel) {
+        result.getMetadata().putToLabels(splitTypeLabel, SPLIT_TYPE_END);
+      }
+      if (null != splitNumberLabel) {
+        result.getMetadata().putToLabels(splitNumberLabel, "1");
+      }
+      splits.add(result);
+      return splits;
+    }
+
     while (idx < n) {
       long timestamp = GTSHelper.tickAtIndex(gts, idx);
       long location = GTSHelper.locationAtIndex(gts, idx);
       long elevation = GTSHelper.elevationAtIndex(gts, idx);
       Object value = GTSHelper.valueAtIndex(gts, idx);
-    
+
+      //
+      // Initialise refLocation as soon as there is a valid position
+      //
+      if (GeoTimeSerie.NO_LOCATION == refLocation && GeoTimeSerie.NO_LOCATION != location) {
+        refLocation = location;
+        refTimestamp = timestamp;
+      }
+
+      //
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
```

Comment:
Can chain the tests in if ... else if ... else if constructs in the order you want (probably the reverse order of the current list of ifs) so we do not execute useless code

Reply:
The most complex one (proximity zone detection) must be done on each iteration. I added a test on proximity radius and time, to save time in case of a distance or time only detection.
Then moved time and distance tests in a if ... else if (with comments to explain priorities)

Unit tests needed before merge.